Add wait counter for nccl abort

atuljangra opened the pull request:
Summary:
Quite a few times, we see the NCCL PG abort taking too long. There's no easy way to measure this, so let's add a counter to measure this across the stack.

This will help us measure how much time we take the NCCL abort. 
Test Plan:
Unit tests

Reviewed By: c-p-i-o

Differential Revision: D62675010


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 commented:
@atuljangra do you mind adding a brief comment above the code describing what the code does? 
Our PR description would be a perfect fit. 
Thanks!

atuljangra commented:
@kwen2501  Added some more details int eh description.

kwen2501 commented:
@atuljangra Thanks. Maybe I confused you :) What I meant was to copy the PR description into the code as a comment (so that readers know what that line of code is for).

atuljangra commented:
@kwen2501 Ah got it, will update.

atuljangra committed:
Add wait counter for nccl abort (#136067)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/136067

Quite a few times, we see the NCCL PG abort taking too long. There's no easy way to measure this, so let's add a counter to measure this across the stack.

Example job where this would have been useful: f641288746-TrainingApplication

Test Plan:
Unit tests

https://www.internalfb.com/mlhub/pipelines/runs/mast/aps-atuljangra-8519d57d57?job_attempt=0&env=PRODUCTION

Reviewed By: c-p-i-o

Differential Revision: D62675010
